File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char
trim :: String -> String
trim xs = reverse (dropWhile isSpace (reverse (dropWhile isSpace xs)))
allSpaces :: String -> Bool
allSpaces xs = all isSpace 
hasUppers :: String -> Bool
hasUppers xs = any isUpper 
allUppers :: String -> Bool
allUppers xs = hasUppers xs && (map toUpper xs == xs)
question :: String -> Bool
question xs = last xs == '?'
exclaim :: String -> Bool
exclaim xs = last xs == '!'

responseFor :: String -> String
responseFor xs | allSpaces trimmed = "Fine. Be that way!"
               | allUppers trimmed && question trimmed = "Calm down, I know what I'm doing!"
               | question trimmed = "Sure."
               | allUppers trimmed && exclaim trimmed = "Whoa, chill out!"
               | allUppers trimmed = "Whoa, chill out!"
               | null trimmed = "Fine. Be that way!"
               | otherwise = "Whatever."  
               where 
                   trimmed = trim xs

File: go/bob/bob.go
package bob

import "strings"
import "unicode"

func hasLetters(remark string) bool {
	for _,rem := range remark{
           if unicode.IsLetter(rem){
		return true;
           }
	}
	return false;
}
func Hey(remark string) string {
	remark = strings.TrimFunc(remark,unicode.IsSpace)
	if remark == ""{
		return "Fine. Be that way!"
	}
	if ! hasLetters(remark) {
	   if strings.HasSuffix(remark,"?"){
               return  "Sure."
           }
	   return "Whatever."
	}
	if remark == strings.ToUpper(remark) && strings.HasSuffix(remark,"?"){
           return "Calm down, I know what I'm doing!"
	}
	if remark == strings.ToUpper(remark) { 
           return "Whoa, chill out!"
	}
	if strings.HasSuffix(remark,"?"){
           return "Sure."
	}
	return "Whatever."
}
